Improve high DPI support docs

vadz opened the pull request:
The dot diagram is still ugly:
![pixels_convert](https://user-images.githubusercontent.com/146917/149260698-0cce1ea8-6522-41c4-8c8c-ab60a5686599.png)
but I just don't know how to position the labels correctly with it.

See #18889.

vadz committed:
Fix code snippets markup in Doxygen overviews

Markdown ``` doesn't work here, Doxygen-specific ~~~ needs to be used.

vadz committed:
Improve documentation of various pixel types

Also add a diagram showing the functions to use to convert between them,
perhaps this can be more clear than textual description.

Co-Authored-By: Stefan Csomor <csomor@advancedconcepts.ch>

vadz committed:
Fix table of contents generation for high DPI overview

Add the required anchors, just the headers are not taken into account
for TOC generation.

vadz committed:
Add wxBitmapBundle section to high DPI overview

Explain how it can be created and used.

csomor commented on docs/doxygen/overviews/high_dpi.md:
```diff
@@ -69,55 +69,194 @@ pixels**, abbreviated as "DIP", that are always of the same size on all
 
-High-Resolution Images and Artwork
+@dot
+digraph Pixels
+{
+    node [shape = hexagon, style = filled];
+
+    LP  [fillcolor = lightblue, label = "Logical\nPixels"];
+    DIP [fillcolor = yellow, label = "DI\nPixels"];
+    PP  [fillcolor = green, label = "Physical\nPixels"];
+
+    LP  -> PP [dir = both, minlen = 3, headlabel = "mul by GetContentScaleFactor()", taillabel = "div by GetContentScaleFactor()"];
+    LP -> DIP [dir = both, weight = 2, minlen = 3, headlabel = "ToDIP()", taillabel = "FromDIP()"];
+    DIP -> PP [dir = both, minlen = 15, headlabel = "mul by GetDPIScaleFactor()", taillabel = "div by GetDPIScaleFactor()", constraint = false] ;
+}
```
how would this look like on your machine:

```
digraph Pixels
{
    node [shape = hexagon, style = filled, fontname = Helvetica];

    LP  [fillcolor = lightblue, label = "Logical\nPixels"];
    DIP [fillcolor = yellow, label = "DI\nPixels"];
    PP  [fillcolor = green, label = "Physical\nPixels"];

    LP  -> PP [fontname = Helvetica, labeldistance = 5, labelangle = 30, dir = both, weight = 2, minlen = 3, label = "GetContentScaleFactor()", headlabel = "multiply by", taillabel = "divide by"];
    LP -> DIP [fontname = Helvetica, labeldistance = 6, dir = both, weight = 2, minlen = 3, headlabel = "ToDIP()", taillabel = "FromDIP()"];
    DIP -> PP [fontname = Helvetica, dir = both, minlen = 10, label = "GetDPIScaleFactor()" headlabel = "multiply by", taillabel = "divide by", constraint = false] ;
}
```

vadz replied:
Much better, thanks!

BTW, I was using [dot visual editor](http://magjac.com/graphviz-visual-editor/) for testing this, I didn't know about it before, but it's really useful.

csomor replied:
> Much better, thanks!
> 
> BTW, I was using [dot visual editor](http://magjac.com/graphviz-visual-editor/) for testing this, I didn't know about it before, but it's really useful.

Thanks, yes, didn't know it either, always great to have something browser-based. I've found the VS-Code extension Graphviz (dot) quite useful as well.

csomor commented on docs/doxygen/overviews/high_dpi.md:
```diff
@@ -69,55 +69,194 @@ pixels**, abbreviated as "DIP", that are always of the same size on all
+wxBitmap or wxImage in wxWidgets API have been updated to work with
+wxBitmapBundle instead, which allows the library to select the appropriate
+size depending on the current DPI and, for the platforms supporting it
+(currently only MSW), even update the bitmap automatically if the DPI changes,
+as can happen, for example, when the window showing the bitmap is moved to
```
actually since we transform a bundle into a NSImage with several representations on macOS, the same happens on macOS automatically as well

vadz replied:
Thanks, changed to "(currently only MSW and macOS)".

csomor commented on docs/doxygen/overviews/high_dpi.md:
```diff
@@ -69,55 +69,194 @@ pixels**, abbreviated as "DIP", that are always of the same size on all
+between different kinds of pixels:
+
+* From DIP to logical pixels: use wxWindow::FromDIP() or wxWindow::ToDIP().
+* Logical pixels and physical pixels: multiply of divide by wxWindow::GetContentScaleFactor().
```
multiply **or** divide

vadz replied:
Thanks for noticing, fixed.

csomor commented on docs/doxygen/overviews/high_dpi.md:
```diff
@@ -69,55 +69,194 @@ pixels**, abbreviated as "DIP", that are always of the same size on all
+
+* From DIP to logical pixels: use wxWindow::FromDIP() or wxWindow::ToDIP().
+* Logical pixels and physical pixels: multiply of divide by wxWindow::GetContentScaleFactor().
+* From DIP to physical pixels: multiply of divide by wxWindow::GetDPIScaleFactor().
```
multiply **or** divide

csomor reviewed (approved):
apart from the small notes, I think it's good. Thanks!

vadz committed:
Document using wxWindow::{From,To}Phys() for LP<->PP conversions

This is simpler and more clear than dividing or multiplying by
GetContentScaleFactor().

vadz committed:
Add a note about rounding the logical pixels up when converting

This might be unexpected, so explicitly warn about possibly not getting
back exactly the same value after a round-trip.

EwoutH commented:
Thanks for this effort, much appreciated!
